src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs

```diff
@@ -95,17 +95,22 @@ private void AddResidualTrivia(SyntaxTriviaList trivia, bool requiresNewLine = f
             {
                 if (requiresNewLine)
                 {
-                    this.AddEndOfLine();
+                    this.AddEndOfLine(GetEndOfLine(trivia));
```

Comment:
unlike anything other, since this one is explicitly requested to add new line, if GetEndOfLine returns no value (since there was no existing end of line on given trivia), I think we should give one here. it could be just searching first end of line it encounter from the root.

Reply:
I believe this is a behavior change so @dotnet/roslyn-compiler team might need to decide? for me, either way seems to be fine to me.

Reply:
> if GetEndOfLine returns no value (since there was no existing end of line on given trivia)

`requiresNewLine` is set to `true` only in [one case](https://github.com/dotnet/roslyn/blob/09b42b8aa2fccdc44c67d93600b4d0c33903cd8f/src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs#L460) when we are dealing with directives and will be called only if the node to be removed [has a directive](https://github.com/dotnet/roslyn/blob/09b42b8aa2fccdc44c67d93600b4d0c33903cd8f/src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs#L415). 
Here is an example when GetEndOfLine will return no value for an endregion directive:
```
[Fact]
public void TestRemoveClassWithBadEndRegionDirective()
  {
    var cu = SyntaxFactory.ParseCompilationUnit("#region A\r\nclass A { } #endregion");
    var m = cu.DescendantNodes().OfType<TypeDeclarationSyntax>().LastOrDefault();
    var cu2 = cu.RemoveNode(m, SyntaxRemoveOptions.KeepEndOfLine | SyntaxRemoveOptions.KeepDirectives);
    var text = cu2.ToFullString();
    Assert.Equal("\r\n#region A\r\n#endregion", text);
  }
```

The first EOL in the result is from the class A and the second EOL is from region directive. For endregion directive no EOL was found so nothing was added. Even if we add a fallback like `?? SyntaxFactory.CarriageReturnLineFeed` the result would be the same as the `AddEndOfLine` method has [an additional check](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs#L106) that prevents from adding an EOL if the previous trivia is also an EOL. So adding a fallback EOL would make no difference as far as I know.

The above test passes also with the old code, so the behavior for this example remains unchanged.

Reply:
In that case, lets add an assert for the result of `GetEndOfLine(trivia)` being non-null. ie:

```suggestion
                    var eol = GetEndOfLine(trivia);
                    Debug.Assert(eol != null);
                    this.AddEndOfLine(eol);
```

Reply:
A possible alternative would be to fall back to `CarriageReturnLineFeed` when we don't have a better idea. I don't have a strong opinion on this.

Tagging @gafter to advise.
The test to illustrate this new behavior is `TestRemoveBadDirectiveWithoutEOL_KeepEndOfLine_KeepDirectives`

Reply:
> lets add an assert for the result of GetEndOfLine(trivia) being non-null

If we add this assert, the above test will fail as no EOL will be found for the `endregion` directive. My point was that we don't need to add an EOL if there is none found, the result in this case will be the same. 
Also, the above case is the only one that will not return an EOL when `requiresNewLine` is `true`, I wasn't able to find any other.

Reply:
From discussion with Neal, it is better to fall back to CRLF (when we can't infer the kind of EOL from the context) than missing the newline entirely.

In other words, something like `this.AddEndOfLine(GetEndOfLine(trivia) ?? CarriageReturnLineFeed);`

Reply:
I've added the fallback, however I am not aware of any case that will actually use it.